### PR TITLE
Fix parallel upload performance issues

### PR DESCRIPTION
**Pull request by harshavardhana**
Tested by transferring 29GiB data, around 246569 files
over 100Gig network.

- master branch
```
time mc mirror /root myminio/testbucket/
real	2m10s
```

- after changes
```
time mc mirror /root myminio/testbucket/
real	1m2s
```

By default runtime.NumCPU() workers are started by default

### CODE REVIEW

**Comment by harshavardhana** [comment]
After testing over 100g network we have some very promising results.

**Comment by kannappanr** [comment]
Some performance numbers uploading 16K files of size 1 KB to AWS
1. Released mc (08/18) `146` seconds
2. This mc - `104` seconds
3. This mc with threads set to `128` - `21` seconds

File: cmd/mirror-main.go [hunk 1]
```diff
@@ -626,13 +622,29 @@ func newMirrorJob(srcURL, dstURL string, isFake, isRemove, isOverwrite, isWatch
...
-		status:         status,
 		statusCh:       make(chan URLs),
-		queueCh:        make(chan func() URLs),
 		wgStatus:       new(sync.WaitGroup),
 		watcher:        NewWatcher(UTCNow()),
 	}
 
+	if wCount > maxParallelWorkers {
+		errorIf(errInvalidArgument().Trace(),
+			"Woker count %d is greater than max workers %d, setting to max %d instead",
```
Comment (kannappanr) [inline cmd/mirror-main.go]: typo, should be `set to max` instead

**Comment by kannappanr** [comment]
Except for the typo and confirmation from AB, that we can get the `worker-count` from user as input, everything looks good. Once we get the confirmation, we can add the same for `cp` command also.

**Comment by harshavardhana** [comment]
This is approved by @abperiasamy we can proceed to review and merge this PR.

**Comment by kannappanr** [comment]
I think based on our tests, for smaller files, around 32 threads make sense. In this case, if someone is using a processor with lots of CPUs, this will work well. For other cases, we will have to see if this brings in a lot of time improvement.

**Comment by harshavardhana** [comment]
> I think based on our tests, for smaller files, around 32 threads make sense. In this case, if someone is using a processor with lots of CPUs, this will work well. For other cases, we will have to see if this brings in a lot of time improvement.

Worker count is increased 2 at a time with 4 secs interval

**Comment by vadmeste** [comment]
@harshavardhana approved, the start number of workers is responsible for the performance enhancement ?

**Comment by harshavardhana** [comment]
Yes @vadmeste and also higher number.

File: cmd/parallel-manager.go [hunk 2]
```diff
@@ -49,16 +51,15 @@ type ParallelManager struct {
...
 		// to create a new one.
 		return
 	}
 
 	// Update number of threads
-	atomic.AddUint32(&p.workersNum, 1)
+	atomic.AddUint32(&p.workersNum, count)
 
 	// Start a new worker
 	p.wg.Add(1)
```
Comment (poornas) [inline cmd/parallel-manager.go]: shouldn't count workers be started here?

File: cmd/parallel-manager.go [hunk 3]
```diff
@@ -119,7 +120,7 @@ func (p *ParallelManager) monitorProgress() {
 					maxBandwidth = bandwidth
 				}
 
-				p.addWorker()
+				p.addWorker(2)
```
Comment (poornas) [inline cmd/parallel-manager.go]: might be better to define a constant, rather than hard code a number?

**Comment by nitisht** [comment]
@harshavardhana can you pls check the appveyor failure

**Comment by harshavardhana** [comment]
Appveyor failure is related to go1.11 upgrade.

**Comment by llamafilm** [comment]
Was anyone able to achieve 10Gbps, like in the linked issue #2366 ?

29GB in 1m2s is only 3.7Gbps.

I'm seeing about 2.5Gbps transferring a single large file over a 20Gb network.

**Pull request merged**